pallets/amm-support/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+// This file is part of hydration-node.
+
+// Copyright (C) 2020-2022  Intergalactic, Limited (GIB).
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::too_many_arguments)]
+
+use crate::types::*;
+use frame_support::sp_runtime::app_crypto::sp_core;
+use frame_support::sp_runtime::{BoundedVec, DispatchError, DispatchResult};
+use frame_system::pallet_prelude::BlockNumberFor;
+use sp_core::ConstU32;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod tests;
+
+pub mod types;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+const LOG_TARGET: &str = "runtime::amm-support";
+
+pub const MAX_STACK_SIZE: u32 = 10;
+
+type ExecutionIdStack = BoundedVec<ExecutionType, ConstU32<MAX_STACK_SIZE>>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::storage]
+	/// Next available incremental ID
+	#[pallet::getter(fn incremental_id)]
+	pub(super) type IncrementalId<T: Config> = StorageValue<_, IncrementalIdType, ValueQuery>;
+
+	#[pallet::storage]
+	/// Execution context to figure out where the trade is originated from
+	#[pallet::getter(fn execution_context)]
+	pub(super) type ExecutionContext<T: Config> = StorageValue<_, ExecutionIdStack, ValueQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Trade executed.
+		Swapped {
+			swapper: T::AccountId,
+			filler: T::AccountId,
+			filler_type: Filler,
+			operation: TradeOperation,
+			inputs: Vec<Asset>,
+			outputs: Vec<Asset>,
+			fees: Vec<Fee<T::AccountId>>,
+			operation_id: Vec<ExecutionType>,
+		},
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			let mut weight: Weight = Weight::zero();
+			weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
```

Comment:
do we read | save anything?

Reply:
Q: Do we need to read or save anything? 

We should have unique block in which something is executed and then execution id can be per block and as such we might wipe it and have 0 reads and writes here? Also not sure to what we configured this but it could be u16 or u32 if we do so

Reply:
as discussed, we leave it as it is for now, then after testing it out by indexer on the test evn, we could do this optimizastion.